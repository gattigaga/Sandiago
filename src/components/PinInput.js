import React, { Component, createRef } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Container = styled.View`
  margin-bottom: 16px;
`;

const InputContainer = styled.View`
  flex-direction: row;
  justify-content: space-between;
`;

export const Input = styled.TextInput`
  width: 48px;
  font-family: Roboto-Bold;
  font-size: 32px;
  color: ${({ color }) => color};
  text-align: center;
`;

export const ErrorText = styled.Text`
  font-family: Roboto-Regular;
  font-size: 12px;
  color: #ff5252;
  text-align: center;
`;

class PinInput extends Component {
  constructor(props) {
    super(props);

    this.state = {
      focusIndex: 0
    };

    this.inputs = props.value.map(createRef);
    this.updateInput = this.updateInput.bind(this);
  }

  /**
   * Handle user input per value part
   *
   * @param {number} index Value part index
   * @returns {function} Callback
   * @memberof PinInput
   */
  updateInput(index) {
    const { value, onChange } = this.props;

    const assign = (inputIndex, newPart) => (_, index) => {
      return inputIndex === index ? newPart : value[index];
    };
    const isValid = index < value.length - 1;

    return newPart => {
      const newValue = value.map(assign(index, newPart));
      onChange(newValue);

      if (isValid) {
        this.inputs[index + 1].current.focus();
      }
    };
  }

  render() {
    const { style, value, isError, errorText } = this.props;
    const { focusIndex } = this.state;

    return (
      <Container style={style}>
        <InputContainer>
          {value.map((_, index) => {
            const ref = this.inputs[index];
            const isFocused = focusIndex === index;
            const currentPart = value[index];
            const focusColor = isFocused ? "#ffb142" : "white";
            const fontColor = isError ? "#ff5252" : focusColor;

            return (
              <Input
                key={index}
                innerRef={ref}
                keyboardType="phone-pad"
                value={currentPart}
                maxLength={1}
                underlineColorAndroid={fontColor}
                color={fontColor}
                onFocus={() => this.setState({ focusIndex: index })}
                onChangeText={this.updateInput(index)}
              />
            );
          })}
        </InputContainer>
        {isError && <ErrorText>{errorText}</ErrorText>}
      </Container>
    );
  }
}

PinInput.propTypes = {
  style: PropTypes.array,
  value: PropTypes.array,
  onChange: PropTypes.func,
  isError: PropTypes.bool,
  errorText: PropTypes.string
};

export default PinInput;
