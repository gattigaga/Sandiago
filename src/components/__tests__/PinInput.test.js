import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import PinInput, { Input, ErrorText, Label } from "../PinInput";

describe("PinInput", () => {
  const setup = propOverrides => {
    const props = {
      label: "Enter your PIN",
      value: [...Array(6)].map(() => ""),
      onChange: jest.fn(),
      ...propOverrides
    };

    const wrapper = shallow(<PinInput {...props} />);

    return {
      wrapper,
      props
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should renders with custom style", () => {
    const style = [{ backgroundColor: "red" }];
    const { wrapper } = setup({ style });

    expect(wrapper.props().style).toEqual(style);
  });

  it("should renders in error state", () => {
    const { wrapper } = setup({ isError: true, errorText: "Pin is required" });
    const $errorText = wrapper.find(ErrorText);
    const $input = wrapper.find(Input);
    const $label = wrapper.find(Label);

    expect($errorText.exists()).toEqual(true);
    expect($label.props().color).toEqual("#ff5252");

    $input.forEach(node => {
      expect(node.props().underlineColorAndroid).toEqual("#ff5252");
      expect(node.props().color).toEqual("#ff5252");
    });
  });
});
