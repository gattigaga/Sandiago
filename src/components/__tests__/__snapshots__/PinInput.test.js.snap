// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinInput should renders default 1`] = `
<Styled(Component)>
  <Styled(Text)
    color="white"
  >
    Enter your PIN
  </Styled(Text)>
  <Styled(Component)>
    <Styled(TextInput)
      color="#ffb142"
      innerRef={
        Object {
          "current": null,
        }
      }
      key="0"
      keyboardType="phone-pad"
      maxLength={1}
      onChangeText={[Function]}
      onFocus={[Function]}
      underlineColorAndroid="#ffb142"
      value=""
    />
    <Styled(TextInput)
      color="white"
      innerRef={
        Object {
          "current": null,
        }
      }
      key="1"
      keyboardType="phone-pad"
      maxLength={1}
      onChangeText={[Function]}
      onFocus={[Function]}
      underlineColorAndroid="white"
      value=""
    />
    <Styled(TextInput)
      color="white"
      innerRef={
        Object {
          "current": null,
        }
      }
      key="2"
      keyboardType="phone-pad"
      maxLength={1}
      onChangeText={[Function]}
      onFocus={[Function]}
      underlineColorAndroid="white"
      value=""
    />
    <Styled(TextInput)
      color="white"
      innerRef={
        Object {
          "current": null,
        }
      }
      key="3"
      keyboardType="phone-pad"
      maxLength={1}
      onChangeText={[Function]}
      onFocus={[Function]}
      underlineColorAndroid="white"
      value=""
    />
    <Styled(TextInput)
      color="white"
      innerRef={
        Object {
          "current": null,
        }
      }
      key="4"
      keyboardType="phone-pad"
      maxLength={1}
      onChangeText={[Function]}
      onFocus={[Function]}
      underlineColorAndroid="white"
      value=""
    />
    <Styled(TextInput)
      color="white"
      innerRef={
        Object {
          "current": null,
        }
      }
      key="5"
      keyboardType="phone-pad"
      maxLength={1}
      onChangeText={[Function]}
      onFocus={[Function]}
      underlineColorAndroid="white"
      value=""
    />
  </Styled(Component)>
</Styled(Component)>
`;
